### PR TITLE
SPI

### PR DESCRIPTION
**Pull request by itchy5**
low level SPI implementation using cubef4

### CODE REVIEW

File: source/cubef4.c [hunk 1]
```diff
@@ -43,4 +43,5 @@ void HAL_IncTick(void)
 void HAL_Delay(uint32_t Delay)
 {
     vTaskDelay(Delay);
+    vTaskDelay(10);
```
Comment (rplauche) [inline source/cubef4.c]: Why is this extra delay needed?

File: source/spi.c [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    hal_spi_handle * handle = NULL;
+    if (spi != NULL)
+    {
+        handle = hal_spi_get_handle(spi->bus_num);
+        if (handle != NULL)
+        {
+            handle->kspi = spi;
+            switch(spi->bus_num)
+            {
+                case K_SPI1:
```
Comment (rplauche) [inline source/spi.c]: I think the STM32F4 has three SPI interfaces. Let's add support for all of them.

File: source/spi.c [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    return handle;
+}
+
+static void hal_spi_terminate(hal_spi_handle * handle)
+{
+    __HAL_RCC_SPI1_FORCE_RESET();
+    __HAL_RCC_SPI1_RELEASE_RESET();
+
+    /* de-init pins */
+    HAL_GPIO_DeInit(GPIOA, GPIO_PIN_5);
```
Comment (rplauche) [inline source/spi.c]: These pins should be pulled from the config data/json, rather than hard coded.

File: source/spi.c [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    if (handle->kspi->bus_num == K_SPI1)
+    {
+        /* Enable SPI clock */
+        __HAL_RCC_SPI1_CLK_ENABLE();
+
+        /* Init pins */
+        hal_spi_gpio_init();
+    }
+        /* Set options */
+        KSPIConf conf = handle->kspi->config;
```
Comment (rplauche) [inline source/spi.c]: Indentation is a bit funky here.

File: source/spi.c [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    return SPI_OK;
+}
+
+
+static void hal_spi_gpio_init()
+{
+    GPIO_InitTypeDef  GPIO_InitStruct;
+    /* SPI GPIO Init */
+    GPIO_InitStruct.Pin       = GPIO_PIN_5 | GPIO_PIN_6 | GPIO_PIN_7;
```
Comment (rplauche) [inline source/spi.c]: Let's pull these pins from the config, rather than hard coding.

File: source/spi.c [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    GPIO_InitStruct.Mode      = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull      = GPIO_NOPULL;
+    GPIO_InitStruct.Speed     = GPIO_SPEED_HIGH;
+    GPIO_InitStruct.Alternate = GPIO_AF5_SPI1;
+    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
```
Comment (rplauche) [inline source/spi.c]: The GPIO port should also come from the config.

File: source/spi.c [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+            SPIHandle->Init.DataSize = SPI_DATASIZE_16BIT;
+        }
+
+        if (conf.clock_phase == K_SPI_CPHA_1EDGE)
+        {
+            SPIHandle->Init.CLKPhase = SPI_PHASE_1EDGE;
+        }
+        else /* 2 edge */
+        {
+            SPIHandle->Init.CLKPhase = SPI_PHASE_1EDGE;
```
Comment (rplauche) [inline source/spi.c]: This should be `SPI_PHASE_2EDGE` yeah?

**Comment by rplauche** [comment]
The function `hal_spi_terminate` needs to be fixed up to use the `handle->pins`, and the appropriate `__HAL_RCC_SPIx_FORCE_RESET/RELEASE_RESET` functions. After that looks good to merge.

**Comment by rplauche** [comment]
Has this code been tested against all three SPI interfaces on the discovery board?

**Comment by itchy5** [comment]
only SPI1

**Comment by rplauche** [comment]
Ok, let's verify again with all three interfaces.

**Comment by itchy5** [comment]
Okay, all 3 interfaces tested to work. doing last commit w/ spi_terminate

**Pull request merged**